### PR TITLE
feature 057

### LINKED ISSUE #57: As a guardian I would like to be able to remove an activity from the week plan so that cancelled activities can be removed from the citizens plan
Tasks:

- [x] When I long press an activity there should be a trashcan icon in the bottom bar that removes the activity from the week plan

### PR DESCRIPTION
**Pull request by LukasRH**
Closes #57 
Solves #187
As a guardian I would like to be able to remove an activity from the week plan so that cancelled activities can be removed from the citizens plan.

qemu-system-x86_64_2019-04-25_16-32-47
qemu-system-x86_64_2019-04-25_16-32-52
qemu-system-x86_64_2019-04-25_16-32-57
qemu-system-x86_64_2019-04-25_16-33-01

### CODE REVIEW

**Comment by User12837985123** [comment]
@aau-giraf/sw602f19 you have been assigned this. Here's your checklist. I marked @Jacob6565 as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [x] Is the code in the right place? (Both in terms of folder structure and class structure)
- [x] Could this code, to the best of your knowledge, have reused existing code?
- [x] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [x] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [x] Are names meaningful and self-documenting?
- [ ] Is it understandable, by reading the code, what it does?
- [x] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [ ] Is it understandable, by reading the tests, what they do?
- [x] Do the tests cover sensible cases? Both happy paths and exception paths?
- [x] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [x] Are error messages in the code understandable?
- [ ] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [x] Are the features that the code claims to implement, actually implemented?
- [x] Does it look like the code has bugs?
- [x] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [x] Are UI messages, that the user might meet, actually helpful?
- [x] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [x] Is there well written code in this? Remember to tell the developer!

**Comment by User12837985123** [comment]
@aau-giraf/sw611f19 you have been assigned this. Here's your checklist. I marked @boginw as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [x] Is the code in the right place? (Both in terms of folder structure and class structure)
- [ ] Could this code, to the best of your knowledge, have reused existing code?
- [ ] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [ ] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [x] Are names meaningful and self-documenting?
- [ ] Is it understandable, by reading the code, what it does?
- [ ] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [x] Is it understandable, by reading the tests, what they do?
- [ ] Do the tests cover sensible cases? Both happy paths and exception paths?
- [ ] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [ ] Are error messages in the code understandable?
- [ ] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [ ] Are the features that the code claims to implement, actually implemented?
- [ ] Does it look like the code has bugs?
- [ ] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [ ] Are UI messages, that the user might meet, actually helpful?
- [ ] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [ ] Is there well written code in this? Remember to tell the developer!

File: lib/blocs/activity_bloc.dart [hunk 1]
```diff
@@ -60,6 +60,7 @@ class ActivityBloc extends BlocBase {
           // A better endpoint would be needed to add the result from the API.
       _activityModelStream.add(_activityModel);
       _weekModel = weekModel;
+      print(weekModel.days[0].activities[0].state);
```
Comment (boginw) [inline lib/blocs/activity_bloc.dart]: Remove this print

File: lib/blocs/toolbar_bloc.dart [hunk 2]
```diff
@@ -20,18 +20,27 @@ class ToolbarBloc extends BlocBase {
   final AuthBloc _authBloc = di.getDependency<AuthBloc>();
 
   //// Based on a list of the enum AppBarIcon this method populates a list of IconButtons to render in the nav-bar
-  void updateIcons(List<AppBarIcon> icons, BuildContext context) {
+  void updateIcons(Map<AppBarIcon, VoidCallback> icons, BuildContext context) {
     List<IconButton> _iconsToAdd;
     _iconsToAdd = <IconButton>[];
 
-    if (icons == null) {
+    icons ??= <AppBarIcon, VoidCallback>{
```
Comment (boginw) [inline lib/blocs/toolbar_bloc.dart:30]: even though this is a very smart operator, one should refrain from it, as it makes the code more difficult for others to read

File: lib/blocs/toolbar_bloc.dart [hunk 3]
```diff
@@ -20,18 +20,27 @@ class ToolbarBloc extends BlocBase {
...
+        AppBarIcon.settings: null,
+        AppBarIcon.logout: null
+      };
+
+    for (AppBarIcon icon in icons.keys) {
+      _addIconButton(_iconsToAdd, icon, icons[icon], context);
+    }
+
+    /*if (icons == null) {
```
Comment (boginw) [inline lib/blocs/toolbar_bloc.dart]: no code comments in source

File: lib/blocs/weekplan_bloc.dart [hunk 4]
```diff
@@ -3,35 +3,94 @@ import 'package:api_client/models/activity_model.dart';
...
+  }
+
+  /// Clears marked activities
+  void clearMarkedActivities() {
+    _markedActivities.add(<ActivityModel>[]);
+  }
+
+  /// Checks if an activity is marked
+  bool isActivityMarked(ActivityModel activityModel) {
+    return _markedActivities.value?.contains(activityModel) ?? false;
```
Comment (boginw) [inline lib/blocs/weekplan_bloc.dart]: again with the operators. Readers will appreciate a more verbose structure

File: lib/screens/weekplan_screen.dart [hunk 5]
```diff
@@ -123,18 +167,90 @@ class WeekplanScreen extends StatelessWidget {
...
+                    },
+                    itemCount: weekday.activities.length + 1,
+                  ),
+                );
+              });
+        });
+  }
+
+  /// Handles tap on a activity
+  void handleOnTapActivity(AsyncSnapshot<bool> snapshot, bool isMarked,
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: This function does not need to take a snapshot of bool, but rather an actual bool.

File: lib/screens/weekplan_screen.dart [hunk 6]
```diff
@@ -201,8 +317,11 @@ class WeekplanScreen extends StatelessWidget {
   }
 
   // Returning a widget that stacks a pictogram and an accept icon
-  FittedBox _pictogramIconStack(
-      BuildContext context, int index, WeekdayModel weekday) {
+  FittedBox _pictogramIconStack(BuildContext context, int index,
+      WeekdayModel weekday, AsyncSnapshot<bool> snapshot) {
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: This function should not use snapshot of bool but rather an actual bool

File: lib/screens/weekplan_screen.dart [hunk 7]
```diff
@@ -164,7 +280,7 @@ class WeekplanScreen extends StatelessWidget {
 
   // Returns the draggable pictograms, which also function as drop targets.
   DragTarget<Tuple2<ActivityModel, Weekday>> _dragTargetPictogram(
-      int index, WeekdayModel weekday) {
+      int index, WeekdayModel weekday, AsyncSnapshot<bool> snapshot) {
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: snapshot of bool is not needed, but rather an actual bool

File: lib/screens/weekplan_screen.dart [hunk 8]
```diff
@@ -254,6 +361,42 @@ class WeekplanScreen extends StatelessWidget {
     );
   }
 
+  /// Builds card that displays the activity
+  Card buildActivityCard(
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: should this be public?

File: lib/blocs/activity_bloc.dart (hunk 1 above)
Comment (User12837985123) [inline lib/blocs/activity_bloc.dart]: Is this a leftover from previous debugging?

File: lib/widgets/giraf_app_bar_widget.dart [hunk 9]
```diff
@@ -16,7 +16,10 @@ class GirafAppBar extends StatelessWidget implements PreferredSizeWidget {
   final String title;
 
   /// Used to store the icons that should be displayed in the appbar.
-  final List<AppBarIcon> appBarIcons;
+  //final List<AppBarIcon> appBarIcons;
```
Comment (boginw) [inline lib/widgets/giraf_app_bar_widget.dart]: commented code

File: lib/widgets/pictogram_image.dart [hunk 10]
```diff
@@ -23,6 +26,9 @@ class PictogramImage extends StatelessWidget {
   /// every press of the image
   final VoidCallback onPressed;
 
+  /// The provided callback function which will be called on
+  /// every long press of the activity
+
```
Comment (boginw) [inline lib/widgets/pictogram_image.dart]: what is this documenting?

File: test/blocs/toolbar_bloc_test.dart [hunk 11]
```diff
@@ -29,10 +29,10 @@ void main() {
...
 
   test('Defined icon is added to stream', async((DoneFn done) {
-    List<AppBarIcon> iconsList;
-    iconsList = <AppBarIcon>[];
-    iconsList.add(AppBarIcon.undo);
-    bloc.updateIcons(iconsList, null);
+    final Map<AppBarIcon, VoidCallback> icons = <AppBarIcon, VoidCallback>
+    {AppBarIcon.undo : null};
+
+    bloc.updateIcons(icons, null);
```
Comment (boginw) [inline test/blocs/toolbar_bloc_test.dart]: update icons should be after listen as to be sure that this is the event we are listening to

File: test/blocs/toolbar_bloc_test.dart [hunk 12]
```diff
@@ -42,11 +42,9 @@ void main() {
...
 
   test('Defined icons are added to stream', async((DoneFn done) {
-    List<AppBarIcon> iconsList;
-    iconsList = <AppBarIcon>[];
-    iconsList.add(AppBarIcon.undo);
-    iconsList.add(AppBarIcon.search);
-    bloc.updateIcons(iconsList, null);
+    final Map<AppBarIcon, VoidCallback> icons = <AppBarIcon, VoidCallback>
+    {AppBarIcon.undo : null, AppBarIcon.search: null};
+    bloc.updateIcons(icons, null);
```
Comment (boginw) [inline test/blocs/toolbar_bloc_test.dart]: update icons should be after listen as to be sure that this is the event we are listening to

File: test/blocs/weekplan_bloc_test.dart [hunk 13]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+            imageHash: null,
+            imageUrl: null,
+            lastEdit: null,
+            title: 'test'),
+        id: 1,
+        isChoiceBoard: null,
+        order: null,
+        state: null);
+
+    weekplanBloc.addMarkedActivity(activityModel);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/toolbar_bloc_test.dart [hunk 14]
```diff
@@ -29,10 +29,10 @@ void main() {
...
-    iconsList = <AppBarIcon>[];
-    iconsList.add(AppBarIcon.undo);
-    bloc.updateIcons(iconsList, null);
+    final Map<AppBarIcon, VoidCallback> icons = <AppBarIcon, VoidCallback>
+    {AppBarIcon.undo : null};
+
+    bloc.updateIcons(icons, null);
 
     bloc.visibleButtons.listen((List<IconButton> response) {
       expect(response.length, 1);
```
Comment (boginw) [inline test/blocs/toolbar_bloc_test.dart:37]: done should be in here!

File: test/blocs/weekplan_bloc_test.dart [hunk 15]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+
+    weekplanBloc.markedActivities
+        .listen((List<ActivityModel> markedActivitiesList) {
+      expect(markedActivitiesList.length, 1);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:100]: this is never executed as the test finishes before this code is reached

File: test/blocs/weekplan_bloc_test.dart [hunk 16]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+        state: null);
+
+    weekplanBloc.addMarkedActivity(firstActivityModel);
+    weekplanBloc.addMarkedActivity(secondActivityModel);
+
+    weekplanBloc.removeMarkedActivity(firstActivityModel);
+
+    weekplanBloc.markedActivities
+        .listen((List<ActivityModel> markedActivitiesList) {
+      expect(markedActivitiesList.length, 1);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: this is never executed as the test finishes before this code is reached

File: test/blocs/weekplan_bloc_test.dart [hunk 17]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+            imageHash: null,
+            imageUrl: null,
+            lastEdit: null,
+            title: 'test123'),
+        id: 2,
+        isChoiceBoard: null,
+        order: null,
+        state: null);
+
+    weekplanBloc.addMarkedActivity(firstActivityModel);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:137]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 18]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+    weekplanBloc.addMarkedActivity(secondActivityModel);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:138]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 19]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+
+    weekplanBloc.removeMarkedActivity(firstActivityModel);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 20]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+            imageHash: null,
+            imageUrl: null,
+            lastEdit: null,
+            title: 'test'),
+        id: 123,
+        isChoiceBoard: null,
+        order: null,
+        state: null));
+
+    weekplanBloc.clearMarkedActivities();
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 21]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+            lastEdit: null,
+            title: 'test123'),
+        id: 2,
+        isChoiceBoard: null,
+        order: null,
+        state: null);
+
+    weekplanBloc.addMarkedActivity(activity);
+
+    weekplanBloc.markedActivities
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 22]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+        .listen((List<ActivityModel> markedActivitiesList) {
+      expect(weekplanBloc.isActivityMarked(activity), true);
+    });
+  });
+
+  test('Checks if the edit mode toggles from false', () {
+    /// Editmode stream initial value is false.
+    weekplanBloc.toggleEditMode();
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 23]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+            lastEdit: null),
+        days: <WeekdayModel>[
+          WeekdayModel(
+              activities: <ActivityModel>[activity], day: Weekday.Monday)
+        ],
+        name: 'Week',
+        weekNumber: 1,
+        weekYear: 2019);
+
+    weekplanBloc.setWeek(weekModel, user);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:242]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 24]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+    weekplanBloc.userWeek.listen((UserWeekModel userWeekModel) {
+      verify(api.week.update(any, any, any, any));
+      expect(userWeekModel.week.days[Weekday.Monday.index].activities,
+          <ActivityModel>[]);
+    });
+  });
+
+  test('Checks if the edit mode toggles from true', () {
+    /// Editmode stream initial value is false.
+    weekplanBloc.toggleEditMode();
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: test/blocs/weekplan_bloc_test.dart [hunk 25]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+    weekplanBloc.toggleEditMode();
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:258]: event should be fired after `listen` to be sure that this is the event that the listener fires on

File: lib/blocs/toolbar_bloc.dart [hunk 26]
```diff
@@ -20,18 +20,27 @@ class ToolbarBloc extends BlocBase {
...
       icons = <AppBarIcon>[];
       icons.add(AppBarIcon.settings);
       icons.add(AppBarIcon.logout);
     }
     for (AppBarIcon icon in icons) {
       _addIconButton(_iconsToAdd, icon, context);
-    }
+    }*/
```
Comment (User12837985123) [inline lib/blocs/toolbar_bloc.dart]: Shouldn't this code be deleted? Otherwise write why not.

File: lib/screens/choose_citizen_screen.dart [hunk 27]
```diff
@@ -34,10 +34,10 @@ class ChooseCitizenScreen extends StatelessWidget {
         child: Dialog(
           child: Scaffold(
             appBar: GirafAppBar(
-              title: 'Vælg borger', 
-              appBarIcons: const <AppBarIcon>[
-                AppBarIcon.logout
-              ], 
+              title: 'Vælg borger',
+              appBarIcons: const <AppBarIcon, VoidCallback>{
+                AppBarIcon.logout: null
```
Comment (User12837985123) [inline lib/screens/choose_citizen_screen.dart:39]: I tested it, and since you already set AppBarIcon.logout to null in toolbar_bloc.dart on line 29, you shouldn't have to do it here. But maybe I missed a corner case.

File: test/blocs/weekplan_bloc_test.dart [hunk 28]
```diff
@@ -55,6 +57,8 @@ void main() {
 
   setUp(() {
     api = Api('any');
+    weekApi = MockWeekApi();
+    api.week = weekApi;
```
Comment (User12837985123) [inline test/blocs/weekplan_bloc_test.dart]: Is this needed considering "api.week = MockWeekApi();" in line 64?

File: test/blocs/weekplan_bloc_test.dart [hunk 29]
```diff
@@ -75,6 +79,175 @@ void main() {
     });
   });
 
+  test('Adds an activity to a list of marked activities', () {
+    final ActivityModel activityModel = ActivityModel(
```
Comment (User12837985123) [inline test/blocs/weekplan_bloc_test.dart:83]: The tests are in general understandable, but this setup part could maybe need a comment. Just describing what is done.

**Review by andpaacoke — DISMISSED** [review]
PO approved 👍 It looks like you also are solving the mark/edit mode issue #59 in this PR

File: test/blocs/weekplan_bloc_test.dart [hunk 30]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+        id: 123,
+        isChoiceBoard: null,
+        order: null,
+        state: null));
+
+    weekplanBloc.clearMarkedActivities();
+
+    weekplanBloc.markedActivities
+        .listen((List<ActivityModel> markedActivitiesList) {
+      expect(markedActivitiesList.length, 0);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:167]: this is never executed as the test finishes before this code is reached

File: test/blocs/weekplan_bloc_test.dart [hunk 31]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+        id: 2,
+        isChoiceBoard: null,
+        order: null,
+        state: null);
+
+    weekplanBloc.addMarkedActivity(activity);
+
+    weekplanBloc.markedActivities
+        .listen((List<ActivityModel> markedActivitiesList) {
+      expect(weekplanBloc.isActivityMarked(activity), true);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:191]: this is never executed as the test finishes before this code is reached

File: test/blocs/weekplan_bloc_test.dart [hunk 32]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+
+    weekplanBloc.setWeek(weekModel, user);
+
+    weekplanBloc.addMarkedActivity(activity);
+
+    weekplanBloc.deleteMarkedActivities();
+
+    weekplanBloc.userWeek.listen((UserWeekModel userWeekModel) {
+      verify(api.week.update(any, any, any, any));
+      expect(userWeekModel.week.days[Weekday.Monday.index].activities,
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:248]: this is never executed as the test finishes before this code is reached

File: test/blocs/weekplan_bloc_test.dart [hunk 33]
```diff
@@ -75,6 +79,175 @@ void main() {
...
+          <ActivityModel>[]);
+    });
+  });
+
+  test('Checks if the edit mode toggles from true', () {
+    /// Editmode stream initial value is false.
+    weekplanBloc.toggleEditMode();
+    weekplanBloc.toggleEditMode();
+
+    weekplanBloc.editMode.listen((bool toggle) {
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: this is never executed as the test finishes before this code is reached

File: test/screens/weekplan_screen_test.dart [hunk 34]
```diff
@@ -112,21 +133,22 @@ void main() {
     await tester.pumpWidget(MaterialApp(home: WeekplanScreen(weekModel, user)));
     await tester.pump();
     await tester.pump();
+    await tester.pump();
```
Comment (User12837985123) [inline test/screens/weekplan_screen_test.dart]: Consider using pumpAndSettle() instead of multiple pumps.

**Review by boginw — CHANGES_REQUESTED** [review]

**Review by User12837985123 — CHANGES_REQUESTED** [review]

File: lib/blocs/weekplan_bloc.dart (hunk 4 above)
Comment (llundbo) [reply lib/blocs/weekplan_bloc.dart]: Replaced with an if statement

File: lib/screens/weekplan_screen.dart (hunk 8 above)
Comment (llundbo) [reply lib/screens/weekplan_screen.dart]: changed to private

File: lib/screens/choose_citizen_screen.dart (hunk 27 above)
Comment (AlsenDev) [reply lib/screens/choose_citizen_screen.dart:39]: In the toolbar_bloc, on line 27-29 the Icons are only set to null if no AppBarIcons is sent to the function. We have made it more readable in a future commit.

**Review by User12837985123 — DISMISSED** [review]

File: lib/blocs/toolbar_bloc.dart (hunk 2 above)
Comment (AlsenDev) [reply lib/blocs/toolbar_bloc.dart:30]: We tried changing it to icons == null, but this causes linter to fail.

image

Comment (boginw) [reply lib/blocs/toolbar_bloc.dart:30]: Alright, maybe we might look into this in the future

File: lib/blocs/weekplan_bloc.dart [hunk 35]
```diff
@@ -3,35 +3,97 @@ import 'package:api_client/models/activity_model.dart';
...
+
+    for (WeekdayModel weekday in week.days) {
+      weekday.activities.removeWhere(
+          (ActivityModel item) => _markedActivities.value.contains(item));
+    }
+
+    clearMarkedActivities();
+    /// Updates the weekplan in the database
+    _api.week.update(user.id, week.weekYear,
+        week.weekNumber, week).listen((WeekModel onData) => null);
```
Comment (boginw) [inline lib/blocs/weekplan_bloc.dart]:
```suggestion
        week.weekNumber, week).listen((WeekModel onData) {});
```

**Review by User12837985123 — DISMISSED** [review]

**Review by boginw — DISMISSED** [review]

**Comment by Alechj** [comment]
After the merge from develop a bug presented itself where the appbar does not have the required buttons. We are looking into this right away.

**Review by User12837985123 — DISMISSED** [review]

**Review by andpaacoke — CHANGES_REQUESTED** [review]
It would be nice if the bottom bar had the same colors as the top bar. It uses linear gradient colors which can also be seen in the design guide

**Review by andpaacoke — DISMISSED** [review]
PO approved 👍

**Review by User12837985123 — DISMISSED** [review]

File: android/app/.classpath [hunk 36]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11/"/>
```
Comment (boginw) [inline android/app/.classpath]: why is this included? Are we sure that all users have eclipse in their classpath?

File: lib/blocs/weekplan_bloc.dart [hunk 37]
```diff
@@ -3,35 +3,97 @@ import 'package:api_client/models/activity_model.dart';
...
+  final BehaviorSubject<bool> _editMode = BehaviorSubject<bool>.seeded(false);
+  final BehaviorSubject<List<ActivityModel>> _markedActivities =
+      BehaviorSubject<List<ActivityModel>>.seeded(<ActivityModel>[]);
   final BehaviorSubject<UserWeekModel> _userWeek =
       BehaviorSubject<UserWeekModel>();
+  final BehaviorSubject<bool> _activityPlaceholderVisible =
+  BehaviorSubject<bool>.seeded(false);
 
-  /// The stream that emits the currently chosen weekplan
+  /// The stream that emits the current chosen weekplan
```
Comment (boginw) [inline lib/blocs/weekplan_bloc.dart]: "currently" is the correct spelling

File: lib/blocs/weekplan_bloc.dart [hunk 38]
```diff
@@ -3,35 +3,97 @@ import 'package:api_client/models/activity_model.dart';
...
+    final WeekModel week = _userWeek.value.week;
+    final UsernameModel user = _userWeek.value.user;
+
+    for (WeekdayModel weekday in week.days) {
+      weekday.activities.removeWhere(
+          (ActivityModel item) => _markedActivities.value.contains(item));
+    }
+
+    clearMarkedActivities();
+    /// Updates the weekplan in the database
```
Comment (boginw) [inline lib/blocs/weekplan_bloc.dart]: no triple-slash-comments in methods. They should only be used for members

File: lib/screens/weekplan_screen.dart [hunk 39]
```diff
@@ -34,15 +36,22 @@ class WeekplanScreen extends StatelessWidget {
     weekplanBloc.setWeek(_week, _user);
   }
 
-  /// The WeekplanBloc that contains the currently chosen week
+  /// The WeekplanBloc that contains the current chosen weekplan
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: "currently" is the correct spelling

File: lib/screens/weekplan_screen.dart [hunk 40]
```diff
@@ -34,15 +36,22 @@ class WeekplanScreen extends StatelessWidget {
...
 
   @override
   Widget build(BuildContext context) {
     return Scaffold(
-      appBar: GirafAppBar(title: 'Ugeplan'),
+      appBar: GirafAppBar(
+        title: 'Ugeplan',
+        appBarIcons: <AppBarIcon, VoidCallback>{
+          AppBarIcon.edit: () => weekplanBloc.toggleEditMode(),
+          AppBarIcon.settings: null,
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: The empty callback should be safer

```suggestion
          AppBarIcon.settings: () {},
```

File: lib/screens/weekplan_screen.dart [hunk 41]
```diff
@@ -34,15 +36,22 @@ class WeekplanScreen extends StatelessWidget {
...
+          AppBarIcon.logout: null
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]:
```suggestion
          AppBarIcon.logout: () {}
```

File: lib/screens/weekplan_screen.dart [hunk 42]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: <Widget>[
+        Expanded(
+          child: Container(
+            decoration: BoxDecoration(
+                gradient: const LinearGradient(
+                    begin: Alignment.topCenter,
+                    end: Alignment.bottomCenter,
+                    stops: <double>[
+                  0.33,
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: 1/3f would be more readable

File: lib/screens/weekplan_screen.dart [hunk 43]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+                  0.66
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: 2/3f would be more readable

File: lib/screens/weekplan_screen.dart [hunk 44]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+                ],
+                    colors: <Color>[
+                  Color.fromRGBO(254, 215, 108, 1),
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart:114]: shouldn't these colors be standardized somewhere? Or aren't the already?

File: lib/screens/weekplan_screen.dart [hunk 45]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+                  Color.fromRGBO(253, 187, 85, 1),
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart:115]: shouldn't these colors be standardized somewhere? Or aren't the already?

File: lib/screens/weekplan_screen.dart [hunk 46]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+                ])),
+            child: IconButton(
+              key: const Key('DeleteActivtiesButton'),
+              iconSize: 50,
+              icon: const Icon(Icons.delete_forever),
+              onPressed: () {
+                /// Shows dialog to confirm/cancel deletion
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: no triple-slash-comments in methods. They should only be used for members

File: lib/screens/weekplan_screen.dart [hunk 47]
```diff
@@ -56,9 +65,80 @@ class WeekplanScreen extends StatelessWidget {
...
+                  weekplanBloc.getNumberOfMarkedActivities().toString() +
+                  ' aktivitet(er)',
+              confirmButtonText: 'Bekræft',
+              confirmButtonIcon:
+                  const ImageIcon(AssetImage('assets/icons/accept.png')),
+              confirmOnPressed: () {
+                weekplanBloc.deleteMarkedActivities();
+                weekplanBloc.toggleEditMode();
+
+                /// Closes the dialog box
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: no triple-slash-comments in methods. They should only be used for members

File: lib/screens/weekplan_screen.dart [hunk 48]
```diff
@@ -123,18 +182,90 @@ class WeekplanScreen extends StatelessWidget {
...
+                            builder: (BuildContext context,
+                                AsyncSnapshot<bool> snapshot) {
+                              return Visibility(
+                                key: const Key('GreyDragVisibleKey'),
+                                visible: snapshot.data,
+                                child: _dragTargetPlaceholder(index, weekday),
+                              );
+                            });
+                      }
+                      return _dragTargetPictogram(
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: for readability use else, even though strictly not necessary.

File: lib/screens/weekplan_screen.dart [hunk 49]
```diff
@@ -253,6 +376,42 @@ class WeekplanScreen extends StatelessWidget {
...
 
+  /// Builds card that displays the activity
+  Card _buildActivityCard(
+    BuildContext context,
+    List<ActivityModel> activities,
+    int index,
+    ActivityState activityState,
+  ) {
+    Widget icon = Container();
+    if (activityState == ActivityState.Completed) {
```
Comment (boginw) [inline lib/screens/weekplan_screen.dart]: this is somewhat confusing to read. I think I'd like a ternary selection more

File: test/blocs/weekplan_bloc_test.dart [hunk 50]
```diff
@@ -75,6 +76,180 @@ void main() {
...
+            imageHash: null,
+            imageUrl: null,
+            lastEdit: null,
+            title: 'test'),
+        id: 1,
+        isChoiceBoard: null,
+        order: null,
+        state: null);
+
+    // Add the ActivityModel to the list of marked activities.
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart]: What are you commenting?

File: test/blocs/weekplan_bloc_test.dart (hunk 16 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart]: this is still not fixed

File: test/blocs/weekplan_bloc_test.dart (hunk 30 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart:167]: this is still not fixed

File: test/blocs/weekplan_bloc_test.dart (hunk 31 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart:191]: still not fixed

File: test/blocs/weekplan_bloc_test.dart (hunk 23 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart:242]: still not fixed

File: test/blocs/weekplan_bloc_test.dart (hunk 32 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart:248]: still not fixed

File: test/blocs/weekplan_bloc_test.dart (hunk 25 above)
Comment (boginw) [reply test/blocs/weekplan_bloc_test.dart:258]: still not fixed

File: test/blocs/weekplan_bloc_test.dart [hunk 51]
```diff
@@ -75,6 +76,180 @@ void main() {
...
+
+    weekplanBloc.deleteMarkedActivities();
+  });
+
+  test('Checks if the edit mode toggles from true', () {
+    /// Edit mode stream initial value is false.
+    weekplanBloc.toggleEditMode();
+
+    weekplanBloc.editMode.skip(1).listen((bool toggle) {
+      expect(toggle, false);
```
Comment (boginw) [inline test/blocs/weekplan_bloc_test.dart:261]: this is never executed as the test finishes before this code is reached

**Review by boginw — DISMISSED** [review]

File: lib/screens/weekplan_screen.dart (hunk 44 above)
Comment (Alechj) [reply lib/screens/weekplan_screen.dart:114]: We agree that it should probably be standardized somewhere as it is currently not. This should probably be a seperate issue however.

File: lib/screens/weekplan_screen.dart (hunk 45 above)
Comment (Alechj) [reply lib/screens/weekplan_screen.dart:115]: Same as above

File: test/blocs/weekplan_bloc_test.dart (hunk 23 above)
Comment (Alechj) [reply test/blocs/weekplan_bloc_test.dart:242]: The event we are testing is fired under the listener, we are testing the weekplanBloc.deleteMarkedActivities(); event and not the weekplanBloc.setWeek(weekModel, user);
We have however added async to this test as we discovered it was missing.

File: lib/screens/weekplan_screen.dart (hunk 44 above)
Comment (boginw) [reply lib/screens/weekplan_screen.dart:114]: If you could create that issue, that would be great

File: lib/screens/weekplan_screen.dart (hunk 45 above)
Comment (boginw) [reply lib/screens/weekplan_screen.dart:115]: If you could create that issue, that would be great

**Review by boginw — APPROVED** [review]

**Review by User12837985123 — APPROVED** [review]

**Comment by AlsenDev** [comment]
Issue regarding colours has been created. #226

**Review by andpaacoke — APPROVED** [review]
